zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -688,17 +737,19 @@ impl UpdateWith<ContextuallyValidBlock> for Chain {
                 ),
             };
 
+            // remove the utxos this produced
+            // uses `tx_by_hash`
+            self.revert_chain_with(&(outputs, transaction_hash, new_outputs), position);
+            // remove the utxos this consumed
+            // uses `tx_by_hash`
+            self.revert_chain_with(&(inputs, transaction_hash, spent_outputs), position);
+
             // remove `transaction.hash` from `tx_by_hash`
             assert!(
                 self.tx_by_hash.remove(transaction_hash).is_some(),
                 "transactions must be present if block was added to chain"
             );
 
-            // remove the utxos this produced
-            self.revert_chain_with(new_outputs, position);
-            // remove the utxos this consumed
-            self.revert_chain_with(inputs, position);
```

Comment:
This was another bug: these methods use `tx_by_hash`, but it was reverted before they ran.